unblob/handlers/archive/xiaomi/hdr.py

```diff
@@ -0,0 +1,197 @@
+          uint64 region;                /* EU */
+          uint64 unused2[2];
+          uint32 blob_offsets[8];
+    } hdr2_header_t;
+
+    struct xiaomi_blob_header {
+          uint32 magic;                 /* 0x0000babe */
+          uint32 flash_offset;
+          uint32 size;                  /* Size of blob */
+          uint16 type;                  /* Type of blob */
+          uint16 unused;
+          char name[32];                /* Name of blob */
+    } blob_header_t;
+
+    struct xiaomi_signature_header {
+          uint32 size;
+          uint32 padding[3];
+          uint8 content[0x100];
+    } signature_header_t;
+ """
+
+
+def calculate_crc(file: File, start_offset: int, size: int) -> int:
+    file.seek(start_offset)
+    content = bytearray(file.read(size))
+    return (binascii.crc32(content) ^ -1) & 0xFFFFFFFF
+
+
+def is_valid_blob_header(blob_header: Instance) -> bool:
+    if blob_header.magic == BLOB_MAGIC:
+        return False
+    if not blob_header.size:
+        return False
+    try:
+        snull(blob_header.name).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.signature_offset < len(header):
+        return True
+    if not header.blob_offsets[0]:
+        return True
+    return True
+
+
+@attrs.define
+class Blob:
+    offset: int
+    name: str
+    size: int
+
+
+class HDRExtractor(Extractor):
+    def __init__(self, header_struct: str):
+        self.header_struct = header_struct
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            blobs = self.parse(file)
+            for blob in blobs:
+                chunk = Chunk(
+                    start_offset=blob.offset, end_offset=blob.offset + blob.size
+                )
+                carve_chunk_to_file(outdir.joinpath(Path(blob.name)), file, chunk)
+
+    def parse(self, file: File) -> List[Blob]:
+        """Parse header + blob headers and populate self._blobs"""
+        blobs = []
+        header = self._struct_parser.parse(self.header_struct, file, Endian.LITTLE)
+        for offset in header.blob_offsets:
+            if not offset:
+                break
+
+            file.seek(offset, io.SEEK_SET)
+            blob_header = self._struct_parser.parse(
+                "blob_header_t", file, Endian.LITTLE
+            )
+            logger.debug("blob_header_t", blob_header_t=blob_header, _verbosity=3)
+            if not is_valid_blob_header(blob_header):
+                raise InvalidInputFormat("Invalid blob header.")
+
+            # file.tell() points to right after the blob_header
+            blobs.append(
+                Blob(
+                    offset=file.tell(),
+                    name=snull(blob_header.name).decode("utf-8"),
+                    size=blob_header.size,
+                )
+            )
```

Comment:
I looked at what we do in our internal handlers, and a good way to limit usage of custom objects is to use `Chunk` object everywhere.

We usually do this by returning a list of tuple like this `[("file_name", Chunk()), ...]`.

You can use that approach here with something like:

```suggestion
            blob_name = 
            blobs.append(
                (
                snull(blob_header.name).decode("utf-8"),
                Chunk(start_offset=file.tell(), end_offset=file.tell()+ blob.size)
            ))
```

Using this approach, you can remove the lines `92-96` and use chunks directly in the `extract` function.